evidence/pool.go

```diff
@@ -92,43 +105,81 @@ func (evpool *Pool) Update(block *types.Block, state sm.State) {
 
 	// remove evidence from pending and mark committed
 	evpool.MarkEvidenceAsCommitted(block.Height, block.Time, block.Evidence.Evidence)
+
+	evpool.cleanupValToLastHeight(block.Height)
 }
 
-// AddEvidence checks the evidence is valid and adds it to the pool.
+// AddEvidence checks the evidence is valid and adds it to the pool. If
+// evidence is composite (ConflictingHeadersEvidence), it will be broken up
+// into smaller pieces.
 func (evpool *Pool) AddEvidence(evidence types.Evidence) error {
+	var (
+		state  = evpool.State()
+		evList = []types.Evidence{evidence}
+	)
 
-	// check if evidence is already stored
-	if evpool.store.Has(evidence) {
-		return ErrEvidenceAlreadyStored{}
+	valSet, err := sm.LoadValidators(evpool.stateDB, evidence.Height())
+	if err != nil {
+		return fmt.Errorf("can't load validators at height #%d: %w", evidence.Height(), err)
 	}
 
-	if err := sm.VerifyEvidence(evpool.stateDB, evpool.State(), evidence); err != nil {
-		return ErrInvalidEvidence{err}
-	}
+	// Break composite evidence into smaller pieces.
+	if ce, ok := evidence.(types.CompositeEvidence); ok {
+		evpool.logger.Info("Breaking up composite evidence", "ev", evidence)
 
-	// fetch the validator and return its voting power as its priority
-	// TODO: something better ?
-	valset, err := sm.LoadValidators(evpool.stateDB, evidence.Height())
-	if err != nil {
-		return err
-	}
-	_, val := valset.GetByAddress(evidence.Address())
-	priority := val.VotingPower
+		blockMeta := evpool.blockStore.LoadBlockMeta(evidence.Height())
+		if blockMeta == nil {
+			return fmt.Errorf("don't have block meta at height #%d", evidence.Height())
+		}
 
-	_, err = evpool.store.AddNewEvidence(evidence, priority)
-	if err != nil {
-		return err
+		if err := ce.VerifyComposite(&blockMeta.Header, valSet); err != nil {
+			return err
+		}
+
+		evList = ce.Split(&blockMeta.Header, valSet, evpool.valToLastHeight)
 	}
 
-	evpool.logger.Info("Verified new evidence of byzantine behaviour", "evidence", evidence)
+	for _, ev := range evList {
+		if evpool.store.Has(evidence) {
+			return ErrEvidenceAlreadyStored{}
+		}
```

Comment:
AFAIK, from the moment of receiving evidence to storing at and even when we propose it, we never run a `ValidateBasic()` on the evidence. I would suggest running the check here just after we've checked whether we have the evidence or not.

Reply:
https://github.com/tendermint/tendermint/pull/4532#discussion_r407422780

we validate evidence in all places we receive it.

Reply:
but maybe I am missing something

Reply:
No you're right. I saw the `msg.ValidateBasic()` and thought that it validated the message itself somehow, not the individual evidences inside them